[wasm] Use response files for `emcc` invocations

radical opened the pull request:
This would avoid running into command max length issues on windows.

Fixes https://github.com/dotnet/runtime/issues/51437 .

```
...
--js-library "C:\Users\Ankit Jain\.nuget\packages\microsoft.netcore.app.runtime.mono.browser-wasm\6.0.0-preview.5.21219.10\runtimes\browser-wasm\native\src\library_mono.js"
--js-library "C:\Users\Ankit Jain\.nuget\packages\microsoft.netcore.app.runtime.mono.browser-wasm\6.0.0-preview.5.21219.10\runtimes\browser-wasm\native\src\pal_random.js"
"C:\dev\w3\project\obj\Debug\net6.0\wasm\System.Private.CoreLib.dll.bc"
"C:\dev\w3\project\obj\Debug\net6.0\wasm\System.Text.Json.dll.bc"
"C:\dev\w3\project\obj\Debug\net6.0\wasm\System.Text.Encodings.Web.dll.bc"
...
```

radical commented:
/cc @pranavkm

lewing commented:
I think you should consider the rsp files as build artifacts and split them up appropriately

radical committed:
[wasm] Use response files for `emcc` invocations

This adds response file for the linker command line, which will be the
longest one. And it uses a known name for the response file.

In future, once we have dependency checking, we can use such response
files for that too.

Fixes #51437 .

radical commented:
> I think you should consider the rsp files as build artifacts and split them up appropriately. I'm not sure a separate task is needed.

- removed the task
- and using a fixed response filename for the linker command line
- for the rest, it can be added when we add dependency checking

radekdoulik reviewed (approved):
LGTM.

I was curious about the multiline response file. Tried to use it with `emcc` on windows and noticed the response file needs to be entered with full path. That is the case here, so I think it should work fine on windows. (`<_EmccLinkerResponseFile>$(_WasmIntermediateOutputPath)emcc-link.rsp</_EmccLinkerResponseFile>` , here the `_WasmIntermediateOutputPath` is normalized msbuild path)

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -486,7 +486,20 @@
+    </PropertyGroup>
+    <WriteLinesToFile Lines="$(_EmccLinkerArguments)" File="$(_EmccLinkerResponseFile)" />
+    <Message Text="Running emcc with $(_EmccLinkerArguments)" Importance="Low" />
+    <Exec Command="emcc @$(_EmccLinkerResponseFile)" EnvironmentVariables="@(EmscriptenEnvVars)" />
```
Super nit, I would consider using `%40` here since `@` has special semantics in MSBuild
```suggestion
    <Exec Command="emcc %40$(_EmccLinkerResponseFile)" EnvironmentVariables="@(EmscriptenEnvVars)" />
```

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -486,7 +486,20 @@
+    <PropertyGroup>
+      <_EmccLinkerArguments>$(EmccLDFlags)</_EmccLinkerArguments>
+      <_EmccLinkerArguments>$(_EmccLinkerArguments) @(_DotnetJSSrcFile->'--js-library &quot;%(Identity)&quot;', ' ')</_EmccLinkerArguments>
+      <_EmccLinkerArguments>$(_EmccLinkerArguments) @(_BitcodeFile->'&quot;%(Identity)&quot;', ' ')</_EmccLinkerArguments>
```
So the nice part of having rsp files and putting args in separate lines is that you technically don't need to quote them. Quoting cli args is hard (https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Private.CoreLib/src/System/PasteArguments.cs) and easy to get wrong in the ordinary case. Less so in specialized cases, but if you could avoid it, it's definitely worth looking in to.

radical committed:
[wasm] Emit emcc arguments, one per line in the rsp file

Having the arguments one per line, also means that we don't have to
quote the file paths. And it is more readable.

radical commented:
Failing test on `runtime (Libraries Test Run release mono Linux x64 Debug) `, and `runtime (Libraries Test Run release mono osx x64 Debug)` is https://github.com/dotnet/runtime/issues/51588

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -486,7 +486,23 @@
+      <_EmccLinkerArguments Include="--js-library %(_DotnetJSSrcFile.Identity)" />
+      <_EmccLinkerArguments Include="@(_BitcodeFile)" />
+      <_EmccLinkerArguments Include="@(_WasmObjects)" />
+      <_EmccLinkerArguments Include="-o $(_WasmIntermediateOutputPath)dotnet.js" />
```
Sorry, I guess their format does require whitespaces to be quoted: https://github.com/emscripten-core/emscripten/blob/main/tools/response_file.py#L16-L53.  Your earlier iteration was more correct.

radical committed:
[wasm] Revert to quote file paths even in response file for emcc

Suggested by Pranavkm:
```
.. their format does require whitespaces to be quoted:
https://github.com/emscripten-core/emscripten/blob/main/tools/response_file.py#L16-L53.
```

radical committed:
Merge remote-tracking branch 'origin/main' into emcc-rsp

lewing reviewed (approved):
Assuming this solves the immediate cmd length issue and the quoting works again this looks ok to me.

radical commented:
> Assuming this solves the immediate cmd length issue and the quoting works again this looks ok to me.

Added snippet of the what the rsp looks like to the description.

```
...
--js-library "C:\Users\Ankit Jain\.nuget\packages\microsoft.netcore.app.runtime.mono.browser-wasm\6.0.0-preview.5.21219.10\runtimes\browser-wasm\native\src\library_mono.js"
--js-library "C:\Users\Ankit Jain\.nuget\packages\microsoft.netcore.app.runtime.mono.browser-wasm\6.0.0-preview.5.21219.10\runtimes\browser-wasm\native\src\pal_random.js"
"C:\dev\w3\project\obj\Debug\net6.0\wasm\System.Private.CoreLib.dll.bc"
"C:\dev\w3\project\obj\Debug\net6.0\wasm\System.Text.Json.dll.bc"
"C:\dev\w3\project\obj\Debug\net6.0\wasm\System.Text.Encodings.Web.dll.bc"
...
```

lewing commented:
/backport to release/6.0-preview4
